Add Mercedes me component

ReneNulschDE opened the pull request:
The `Mercedes me` component offers integration with the [Mercedes me](https://www.mercedes-benz.com/de/mercedes-me/) cloud service and provides presence detection as well as sensors such as doors, tires, windows, and service interval.

This component provides the following platforms:
 - Binary Sensors - such as windows, tires, doors, lock
 - Sensors - such as fuel status, service interval, remaining km...
 - Device tracker - to track location of your car

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4464

```yaml
# Example configuration.yaml entry
mercedesme:
  username: email
  password: password
```


## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

ReneNulschDE committed:
Add Mercedes me component

ReneNulschDE committed:
Add Mercedes me component

ReneNulschDE committed:
pump api version to 0.1.2

ReneNulschDE committed:
pump api version to 0.1.2

ReneNulschDE committed:
Merge branch 'mercedesme' of https://github.com/ReneNulschDE/home-assistant into mercedesme

ReneNulschDE committed:
clean up code

ReneNulschDE committed:
Code cleanup

ReneNulschDE committed:
Remove unneeded return statement

ReneNulschDE committed:
Return statements added again

armills commented on homeassistant/components/mercedesme.py:
```diff
@@ -0,0 +1,73 @@
+        }
+    except Exceptions.MercedesMeException as ex:
+        if ex.code == 401:
+            hass.components.persistent_notification.create(
```
I don't really like that we're using persistent notifications to send errors to the user, but it looks like we've got other platforms already doing this. I'll let someone else make the call here.

At the least we should include this same message in `_LOGGER.error`.

armills commented on homeassistant/components/mercedesme.py:
```diff
@@ -0,0 +1,73 @@
+    try:
+        hass.data[DATA_MME] = {
+            'controller': controller.Controller(
+                config[DATA_MME][CONF_USERNAME],
```
These should be `config[DOMAIN]` instead of `config[DATA_MME]`.

armills commented on homeassistant/components/mercedesme.py:
```diff
@@ -0,0 +1,73 @@
+                      ex.message)
+        return False
+
+    if hass.data[DATA_MME]["controller"].is_valid_session:
```
Let's change this to 
```python
if not hass.data[DATA_MME]["controller"].is_valid_session:
    # Log error
    return False
```

If we aren't going to load any entities we should fail the setup.

armills commented on homeassistant/components/mercedesme.py:
```diff
@@ -0,0 +1,73 @@
+        return False
+
+    if hass.data[DATA_MME]["controller"].is_valid_session:
+        discovery.load_platform(hass, 'sensor', DATA_MME, {}, config)
```
Change `DATA_MME` to `DOMAIN` here as well.

armills commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,109 @@
+
+DEPENDENCIES = ['mercedesme']
+
+DATA_MME = 'mercedesme'
```
Change this to an import above:
```python
from homeassistant.components.mercedesme import DATA_MME
```

armills commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,109 @@
+        add_devices([BinarySensor('tireWarningLight', car, controller)], True)
+
+
+class BinarySensor(BinarySensorDevice):
```
We generally include the platform name in the entity class names. Maybe `MercedesMEBinarySensor`?

armills commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,109 @@
+        add_devices([BinarySensor('doorsClosed', car, controller)], True)
+        add_devices([BinarySensor('windowsClosed', car, controller)], True)
+        add_devices([BinarySensor('locked', car, controller)], True)
+        add_devices([BinarySensor('tireWarningLight', car, controller)], True)
```
Collect all the devices that should be added into one list and make one call to `add_devices`.

armills commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,109 @@
+class BinarySensor(BinarySensorDevice):
+    """Representation of a Sensor."""
+
+    def __init__(self, SensorName, Car, Controller):
```
Change these to snake case. Not actually sure how that went through the linter.
```python
def __init__(self, sensor_name, car, controller):
```

armills commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,109 @@
+        self._state = None
+        self._name = SensorName
+        self._sensor_type = None
+        self.__car = Car
```
One underscore is enough: `self._car`.

armills commented on homeassistant/components/device_tracker/mercedesme.py:
```diff
@@ -0,0 +1,62 @@
+        self.controller = controller
+        self.update_info()
+
+        track_utc_time_change(
```
Instead of registering the update yourself, inherit from `device_tracker.DeviceScanner`, and implement `get_scanner` instead of `setup_scanner`.

Example: https://github.com/home-assistant/home-assistant/blob/46e030662da4ab7c653a23e1f34e527c9a0e7a8f/homeassistant/components/device_tracker/bbox.py

ReneNulschDE replied:
Thank you @armills for the review. I've taken your points and changed the code. Only this comment takes longer. All current implementations that are using the `device_tracker.DeviceScanner` class have not implemented GPS coordinates. Could it be that the DeviceScanner is for local network devices and the DeviceTracker for devices that are not directly connected to the local network? All other car components using `setup_scanner`. If it's ok for you, I'll check, learn and implement this in the next cycle. thx

armills replied:
Nice work with the follow up!

I'm OK holding off on the larger refactor, but we should at least switch to `track_time_interval` with the configured time interval. There's an example of an invocation below. It should be a pretty quick change.

https://github.com/home-assistant/home-assistant/blob/93c8372c30200071d54a9e1710805d2bc91688a3/homeassistant/components/raincloud.py#L123

armills commented on homeassistant/components/sensor/mercedesme.py:
```diff
@@ -0,0 +1,98 @@
+class Sensor(Entity):
+    """Representation of a Sensor."""
+
+    def __init__(self, SensorName, Car, Controller, Unit):
```
Same here. Use snake case and prefix with only one underscore for private attributes.

armills commented on homeassistant/components/device_tracker/mercedesme.py:
```diff
@@ -0,0 +1,62 @@
+"""
+import logging
+from homeassistant.helpers.event import track_utc_time_change
+
```
`DEPENDENCIES = ['mercedesme']` is missing in the device_tracker.

ReneNulschDE committed:
Implement requested changes

ReneNulschDE committed:
Rework component, move data load to component

ReneNulschDE committed:
lint

ReneNulschDE committed:
remove debug logging

ReneNulschDE committed:
Change RainCloud comments, change from track_utc_time to track_time_interval

ReneNulschDE commented:
@armills : I have implemented some major changes. Now the component is responsible for the update of the sensors. Only the device tracker is not following this structure right now, but I changed the code based on your comments. May I ask you for review. Thank you.

armills commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,107 @@
+        return retval
+
+    @property
+    def state(self):
```
`BinarySensorDevice` already defines `state`. We can get rid of this whole function definition.

armills commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,107 @@
+                    self._car["lastUpdate"]
+                    ).strftime('%Y-%m-%d %H:%M:%S'),
+                "car": self._car["license"],
+                "Is_On": self.is_on,
```
This is already the binary sensor state and isn't necessary to include in the attributes.

armills commented on homeassistant/components/mercedesme.py:
```diff
@@ -0,0 +1,149 @@
+        self.schedule_update_ha_state(True)
+
+    @property
+    def icon(self):
```
This property can be removed.

ReneNulschDE replied:
Once again, thank you for your review and the more than valid change requests. I've updated the code. Hope that we can land this in the next version. After that, I'll start to add new features.

emlove reviewed (changes requested):
Just a few small follow ups, and this is good to go!

ReneNulschDE committed:
Final cleanup for version 1

emlove reviewed (approved):
This LGTM. Nice work!

armills commented:
Thanks for the contribution! 🌟

MartinHjelmare commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,100 @@
+        """Return the state attributes."""
+        if self._name == "windowsClosed":
+            return {
+                "windowStatusFrontLeft": self._car["windowStatusFrontLeft"],
```
State attributes should be lowercase snakecase.

MartinHjelmare commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,100 @@
+        else:
+            result = self._car[self._name] is True
+
+        self._state = "On" if result else "Off"
```
Why store a string instead of a boolean?

MartinHjelmare commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,100 @@
+    data = hass.data[DATA_MME].data
+
+    if not data.cars:
+        _LOGGER.error("setup_platform data.cars is none")
```
The check is not if data.cars is `None`.

MartinHjelmare commented on homeassistant/components/device_tracker/mercedesme.py:
```diff
@@ -0,0 +1,72 @@
+
+    def __init__(self, hass, config, see, data):
+        """Initialize the Mercedes ME device tracker."""
+        self.hass = hass
```
This attribute isn't needed in any other method.

MartinHjelmare commented on homeassistant/components/mercedesme.py:
```diff
@@ -0,0 +1,144 @@
+        async_dispatcher_connect(
+            self.hass, SIGNAL_UPDATE_MERCEDESME, self._update_callback)
+
+    def _update_callback(self):
```
This could be made a `core.callback` via annotating the method with that.

MartinHjelmare commented on homeassistant/components/mercedesme.py:
```diff
@@ -0,0 +1,144 @@
+
+    def _update_callback(self):
+        """Callback update method."""
+        self.schedule_update_ha_state(True)
```
If the method is made a callback this should be changed to the async version.

MartinHjelmare commented on homeassistant/components/sensor/mercedesme.py:
```diff
@@ -0,0 +1,92 @@
+        """Return the state attributes."""
+        if self._internal_name == "latestTrip":
+            return {
+                "durationSeconds":
```
State attributes should be lowercase snakecase.

ReneNulschDE replied:
Thank you for the feedback. I'll start the rework and will provide a PR in the next version.

MartinHjelmare commented on homeassistant/components/binary_sensor/mercedesme.py:
```diff
@@ -0,0 +1,100 @@
+Support for Mercedes cars with Mercedes ME.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/mercedesme/
```
Wrong url, see other platforms for correct form.

MartinHjelmare reviewed (commented):
Left comments where code could be improved in a future PR.

armills commented:
Oof. Thanks for the follow-up @MartinHjelmare. Sorry for the premature merge!

elRadix commented:
can you please check the issue reported here:
https://community.home-assistant.io/t/0-63-entity-registry-sql-sensor-mercedes-cars/42886/4?u=elradix
